src/js/utils/can-autoplay.js

```diff
@@ -0,0 +1,94 @@
+
+function startPlayback (element, { timeout }) {
+    element.src = VIDEO;
+
+    return new Promise((resolve, reject) => {
+        const playResult = element.play() || createPlayPromise(element);
+        const timeoutId = setTimeout(() => {
+            reject(new Error(`Timeout ${timeout} ms has been reached`));
+        }, timeout);
+        const sendOutput = (result) => {
+            clearTimeout(timeoutId);
+            resolve(result);
+        };
+
+        playResult
+            .then(() => sendOutput(true))
+            .catch(() => sendOutput(false));
+    });
+}
+
+export const AUTOPLAY_ENABLED = 'autoplayEnabled';
+export const AUTOPLAY_MUTED = 'autoplayMuted';
+export const AUTOPLAY_DISABLED = 'autoplayDisabled';
+
+export function canAutoplay (controller, { muted = false, allowMuted = false, timeout = 250 }) {
+    // Set-up.
+    const mediaPool = controller.mediaPool;
+    const element = mediaPool.getPrimedElement();
+    const initialMute = element.muted;
+    const reset = () => {
+        mediaPool.syncMute(initialMute);
+        mediaPool.recycle(element);
+        controller.attached = true; // Re-attach.
+    };
+    controller.attached = false; // Detach.
```

Comment:
I don't think we should detach and attach controllers. That functionality is specific to switching binding of events from the playlist item to something else, but this test should not interfere with that. Calling `attached = true` might do more bad than good.

What we will need to do handle any change in state that might have happened while the test was being performed. Example: user clicks to start playback while test is running. This could be done externally by making the `then` callback use our `cancelable` util.

Reply:
The test does interfere with the model if there is only one `<video>` element - it seems that events on the video element trigger model changes without detaching. Is there a better way to reset the model after the test, or ideally, detach the video event listeners during the test?

Regarding cancellation, yes, we need to handle this, which `then` callback are you referring to? I see trouble ahead when canceling right after `startPlayback` - the `then` will not be executed, but the attributes of the `<video>` element will still be altered in `startPlayback`, which conflicts when the user clicked to start playback. What would help here is if we can attach an `onCancel` handler to the promise somehow, so we can clean things up.

Reply:
@vseventer In this case we must use a different video tag than one attached to the player. The test would wipe out any preloading performed on iOS or Safari Mac and we don't want that. Maybe we can extend the pool to provide a primed tag for this test that isn't the main "mediaElement".